app/models/rental.rb

```diff
@@ -92,34 +110,77 @@ def renter_is_assignable
+  def create_reservations
+    # if we fail half way through creating reservations we should roll them all back
+    created_reservations = []
     begin
-      reservation = Inventory.create_reservation(item_type.name, start_time, end_time)
-      raise 'Reservation UUID was not present in response.' unless reservation[:uuid].present?
+      raise RecordInvalid, 'Rental invalid' unless valid?
+      rentals_items.each do |ri|
+        raise RecordInvalid, 'Rental item is invalid' unless ri.valid? # check if the current rental item is valid
+
+        reservation = Inventory.create_reservation(ri.item_type.name, start_time, end_time)
+        raise 'Reservation UUID was not present in response.' unless reservation[:uuid].present?
 
-      self.reservation_id = reservation[:uuid]
-      self.item = Item.find_by(name: reservation[:item][:name])
+        ri.reservation_id = reservation[:uuid]
+        created_reservations << reservation[:uuid]
+        ri.item = Item.find_by(name: reservation[:item][:name])
+      end
     rescue => error
-      errors.add :base, error.inspect
+      # roll back the rentals we got partially through creating
+      failed_roll_back = false
+      created_reservations.each do |uuid|
+        # remove reservation from RentalsItem
+        if (ri = rentals_items.to_a.find { |_ri| _ri.reservation_id == uuid })
+          ri.reservation_id = nil
+        else
+          failed_roll_back = true
+          errors.add :base, "Failed to find associated rentals item when rolling back reservation (uuid #{uuid})"
+        end
+
+        # remove reservation from api
+        unless delete_reservation(uuid)
+          errors.add :base, "Failed to delete reservations from inventory api (uuid #{uuid})"
+          failed_roll_back = true
+        end
+      end
+
+      errors.add :base, "Reservations #{'partially' if failed_roll_back} rolled back " + error.inspect
       throw :abort
     end
   end
 
-  def delete_reservation
-    return true if reservation_id.nil? # nothing to delete here
+  def delete_reservations
     return true if end_time < Time.current # deleting it is pointless, it wont inhibit new rentals and it will destroy a record.
-    begin
-      Inventory.delete_reservation(reservation_id)
-      self.reservation_id = nil
-    rescue => error
-      errors.add(:base, error.inspect) && (return false)
+    rentals_items.each do |ri|
+      next if ri.reservation_id.nil? # nothing to delete here
+      errors.add(:rentals_items, "Failed to delete reservation (uuid #{ri.reservation_id})") unless delete_reservation(ri.reservation_id)
+      ri.reservation_id = nil
     end
-    true
+    throw(:abort) if errors.any?
+    errors.empty?
+  end
+
+  def delete_reservation(uuid)
+    Inventory.delete_reservation(uuid) # will throw errors if it fails
+  rescue
+    return false
+  end
+
+  def str_reservation_ids
+    rentals_items.reduce('') { |acc, part| "#{acc}, #{part.reservation_id}" }[2..-1]
```

Comment:
This (and the two methods below it) definitely need comments that give an example of what they do.